pkg/agent/openflow/client.go

```diff
@@ -312,6 +313,44 @@ func (c *client) addFlows(cache *flowCategoryCache, flowCacheKey string, flows [
 	return nil
 }
 
+// modifyFlows set the flows of flowCategoryCache be exactly same as the input.
+func (c *client) modifyFlows(cache *flowCategoryCache, flowCacheKey string, flows []binding.Flow) error {
+	oldFlowCacheI, ok := cache.Load(flowCacheKey)
+	var err error
+	if !ok {
+		err = c.ofEntryOperations.AddAll(flows)
+	} else {
+		var adds, mods, dels []binding.Flow
+		oldFlowCache := oldFlowCacheI.(flowCache)
+		newFlowsMap := map[string]bool{}
+		for _, flow := range flows {
+			if _, ok := oldFlowCache[flow.MatchString()]; ok {
+				mods = append(mods, flow)
+			} else {
+				adds = append(adds, flow)
+			}
+			newFlowsMap[flow.MatchString()] = true
```

Comment:
In the modify case, `MatchString()` for each flow is calculated 3 times. Consider the following code to avoid the overhead and an extra map:
```
fCache := flowCache{}
for _, flow := range flows {
        matchString := flow.MatchString()
        if _, ok := oldFlowCache[matchString]; ok {
                mods = append(mods, flow)
        } else {
                adds = append(adds, flow)
        }
        fCache[matchString] = flow
        for k, v := range oldFlowCache {
                if _, ok := fCache[k]; !ok {
                        dels = append(dels, v)
                }
        }
}
```

Reply:
Yes. I created the variable `flow` to store value of `flow.MatchString()`and reused `fCache` instead of `newFlowsMap`. But I guess the for loop of `oldFlowCache` is put in the for loop of `flows` by mistake?

Reply:
Yes, that's a typo